rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,514 @@
+
+```rust
+let circle_mesh: Mesh = Circle2d{ radius: 2.0 }.mesh();
+```
+The base primitive types only define the shape (`Circle2d`) and size (`radius`) of the geometry about the origin. Once generated, the mesh can have a transform applied to it like any other mesh, and it is no longer tied to the primitive that generated it. The `Default::default()` implementation of primitives should be a "unit" variant, such as a circle with diameter 1.0, or a box with all edges of length 1.0.
+
+Meshing could be naturally extended with other libraries or parameters. For example, a sphere by default might use `Icosphere`, but could be extended to generate UV spheres or quad spheres. For 2d types, we can use a crate such as `lyon` to generate 2d meshes from parameterized primitive shapes within the `Meshable` interface.
+
+### Bounding
+
+For use in spatial queries, broad phase collision detection, and raycasting, `Bounding` and `Bounding2d` traits are provided. These traits are implemented for types that define position in addition to the shape and size defined in the base primitive. The basic functionality of this trait is to check whether one bounding shape is contained within another bounding shape.
+
+### Colliders
+
+Colliders can provide intersection checks against other colliders, as well as check if they are within a bounding volume.
+
+### 3D and 2D
+
+This RFC provides independent 2d and 3d primitives. Recall that the purpose of this is to provide lightweight types, so there are what appear to be duplicates in 2d and 3d, such as `Line` and `Line2d`. Note that the 2d version of a line is smaller than its 3d counterpart because it is only defined in 2d. 3d geometry (or 2d with depth) is assumed to be the default for most cases. The names of the types were chosen with this in mind.
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
+```
+### Traits
+
+These traits are provided as reference to illustrate how these might be used. The details of implementation and interface should be determined in a separate RFC, PR, or independent prototypes.
+
+```rust
+trait Meshable{
+  fn mesh(&self) -> Mesh;
+};
+
+trait Bounding {
+  fn within(&self, other: &impl Bounding) -> bool;
+  fn contains(&self, collider: &impl Collider) -> bool;
+}
+
+trait Bounding2d {
+  fn within(&self, other: &impl Bounding2d) -> bool;
+  fn contains(&self, collider: &impl Collider2d) -> bool;
+}
+
+trait Collider {
+  fn collide(&self, other: &impl Collider) -> Option(Collision);
+  fn within(&self, bounds: &impl Bounding) -> bool;
+}
+
+trait Collider2d {
+  fn collide(&self, other: &impl Collider2d) -> Option(Collision2d);
+  fn within(&self, bounds: &impl Bounding2d) -> bool;
+}
+```
+
+### 3D Geometry Types
+
+```rust
+struct Point(Vec3)
```

Comment:
Would this be a pub newtype or would it implement its own methods?

Reply:
I would hope that we can implement methods on it as needed. Are there reasons not to want that?

Reply:
Yes, one reason for wrapping the Vec3 is to implement its own methods in future. The primary reason is to apply type constraints to primitive shape fields to prevent errors such as using a direction (Vec3) instead of a point (Vec3).

Reply:
I would guess it would be both -- it's not maintaining an invariant, so (unlike `Direction`) there's no need to *prevent* people from getting the `.0`, but I'd expect that it would have its own methods and operators and such.  (Including the obvious `Point: Add<Vec, Output = Point>` and `Point: Sub<Point, Output = Vec>`, but also things like `impl !Add<Point> for Point {}`, once arbitrary negative impls are available on stable.)

Reply:
Negative impls sound *great* for this RFC in general.

Reply:
If you want a head start on thinking about this there's a point RFC for glam https://github.com/bitshifter/glam-rs/issues/133.  I'm not currently intending to add point types to glam but the RFC might be of interest here.

Reply:
IMO, the strongest argument here is for type safety / code correctness reasons. Points and vectors can both be stored as a `Vec3`, but mixing them up by passing in the wrong version to an API is very easy to do and tends to result in very annoying to track bugs.

I would make the same argument about coordinate systems, but that's a bridge too far in the context of games :p

Reply:
It's worth noting that points *transform differently* than vectors, since translation should never be applied to a vector. IMO, this is a good reason to make a type-level distinction, support homogeneous coordinates, or both. But I agree with cart that this discussion is orthogonal to this RFC.